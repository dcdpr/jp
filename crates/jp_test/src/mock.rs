use std::{
    env, fs,
    path::{Path, PathBuf},
};

use httpmock::{MockServer, RecordingRuleBuilder};
use saphyr::{LoadableYamlNode, Yaml, YamlEmitter};

/// A recorder for HTTP requests.
pub struct Vcr {
    forward_to: String,
    fixtures: PathBuf,
    recording: bool,
}

impl Vcr {
    /// Create a new recorder.
    pub fn new(forward_to: impl Into<String>, fixtures: impl Into<PathBuf>) -> Self {
        Self {
            forward_to: forward_to.into(),
            fixtures: fixtures.into(),
            recording: env::var("RECORD").is_ok(),
        }
    }

    /// Enable recording for the current test.
    pub fn record(&mut self) {
        self.recording = true;
    }

    /// Enable playback for the current test.
    pub fn playback(&mut self) {
        self.recording = false;
    }

    /// Set recording mode for the current test.
    pub fn set_recording(&mut self, recording: bool) {
        self.recording = recording;
    }

    /// Record a test cassette.
    pub async fn cassette<R, F, Fut, O>(
        &self,
        name: &str,
        rule_builder: R,
        test_fn: F,
    ) -> Result<(), Box<dyn std::error::Error>>
    where
        R: FnOnce(RecordingRuleBuilder),
        F: FnOnce(bool, String) -> Fut,
        Fut: Future<Output = O>,
        O: std::fmt::Debug,
    {
        let fixture = self.fixtures.join(format!("{name}.yml"));
        let server = MockServer::start_async().await;

        if self.recording {
            server
                .forward_to_async(&self.forward_to, |rule| {
                    rule.filter(|when| {
                        when.any_request();
                    });
                })
                .await;

            let recording = server.record_async(rule_builder).await;

            let out = test_fn(true, server.base_url()).await;
            self.verify(name, out);

            let temp_path = recording.save_to_async(&self.fixtures, name).await?;

            if temp_path != fixture && temp_path.exists() {
                fs::rename(temp_path, &fixture)?;
            }

            modify_fixture(&fixture)?;
        } else if fixture.exists() {
            server.playback_async(&fixture).await;
            let out = test_fn(false, server.base_url()).await;
            self.verify(name, out);
        } else {
            return Err(format!("Recording not found at {}", fixture.display()).into());
        }

        Ok(())
    }

    fn verify<T: std::fmt::Debug>(&self, name: &str, expr: T) {
        insta::with_settings!({ snapshot_path => &self.fixtures, prepend_module_to_snapshot => false }, {
            insta::assert_debug_snapshot!(name, expr);
        });
    }
}

fn modify_fixture(fixture: &Path) -> Result<(), Box<dyn std::error::Error>> {
    let contents = fs::read_to_string(fixture)?;
    let mut document = Yaml::load_from_str(&contents)?;
    for yaml in &mut document {
        let Some(header) = yaml
            .as_mapping_get_mut("then")
            .and_then(|then| then.as_mapping_get_mut("header"))
            .and_then(|header| header.as_sequence_mut())
        else {
            return Ok(());
        };

        header.retain(|header| {
            header.as_mapping_get("name").and_then(|v| v.as_str()) == Some("content-type")
        });
    }

    let mut buf = String::new();
    for yaml in &document {
        YamlEmitter::new(&mut buf).dump(yaml)?;
        buf.push('\n');
    }
    fs::write(fixture, buf.trim_start_matches("---\n"))?;

    Ok(())
}
