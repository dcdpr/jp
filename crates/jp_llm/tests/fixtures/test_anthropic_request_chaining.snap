---
source: crates/jp_test/src/mock.rs
expression: expr
---
[
    Ok(
        ChatChunk(
            Content(
                "Kirigami-Inspired Parachutes: Revolutionizing Aerial Design through Origami-Like Techniques\n\nIntroduction\n\nThe world of aerospace engineering and design is constantly evolving,",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                " with researchers and innovators seeking novel approaches to solve complex challenges.",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                " One of the most fascinating recent developments in this field is the emergence of kirigami-inspired parachute designs. Drawing from the ancient Japanese art of paper cutting and folding,",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                " these innovative parachute concepts are reshaping our understanding of aerial deployment, payload delivery, and scientific exploration.\n\nUnderstanding Kirigami and Its Principles\n\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "To appreciate the significance of kirigami-inspired parachutes, it's essential to first understand what kirigami is. Unlike traditional origami, which focuses solely on folding paper,",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                " kirigami involves both cutting and folding. This technique allows for more complex and dynamic transformations of a two-dimensional surface into three-dimensional structures.",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                " The word \"kirigami\" comes from the Japanese words \"kiri\" (cutting) and \"kami\" (paper), reflecting the method's core technique.\n\nThe key principles of kirigami include:\n1.",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                " Strategic cutting of a base material\n2. Precise folding mechanisms\n3. Ability to create complex three-dimensional structures from a flat surface\n4.",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                " Maximizing structural integrity through carefully designed cuts and folds\n\nThese principles translate remarkably well to parachute design,",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                " offering unique advantages over traditional parachute manufacturing approaches.\n\nThe Evolution of Parachute Design\n\nTraditional parachutes have remained relatively unchanged for decades.",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                " Typically constructed from lightweight fabrics and featuring a standard canopy design, they have served critical functions in military, scientific, and recreational applications. However,",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                " these conventional designs often face limitations in terms of packability, deployment efficiency, and adaptability to different environmental conditions.\n\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "Kirigami-inspired approaches introduce a paradigm shift in parachute engineering by reimagining how these aerial devices can be structured, deployed, and utilized.\n\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "Key Innovations in Kirigami-Inspired Parachute Design\n\n1. Compact Deployment Mechanisms\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "One of the most significant advantages of kirigami-inspired parachutes is their ability to be extremely compact when folded. By incorporating strategic cuts and folding patterns,",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                " these parachutes can be compressed into remarkably small volumes, making them ideal for space missions, drone delivery systems, and emergency rescue operations.\n\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "Researchers have developed parachute designs that can reduce the initial volume by up to 90% compared to traditional parachute systems.",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                " This compression is achieved through carefully calculated cut patterns that allow for precise folding without compromising structural integrity.\n\n2. Enhanced Aerodynamic Performance\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "The intricate cutting and folding techniques enable kirigami-inspired parachutes to have more sophisticated aerodynamic characteristics. By modifying the surface geometry through precise cuts,",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                " engineers can create parachutes with:\n- Improved stability during descent\n- Better control over descent rates\n- Reduced oscillation\n- Enhanced wind resistance\n\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "These performance improvements are particularly valuable in scientific applications, such as planetary exploration missions or atmospheric research.\n\n3. Adaptive Morphology\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "Perhaps the most revolutionary aspect of kirigami-inspired parachutes is their ability to change shape dynamically during descent. Unlike traditional parachutes with fixed geometries,",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                " these advanced designs can modify their configuration in response to environmental conditions.\n\nFor example, a kirigami parachute might:\n- Adjust its surface area based on wind speed\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "- Change its shape to optimize drag\n- Redistribute structural stress during descent\n- Automatically compensate for atmospheric variations\n\n4. Material Innovations\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "Kirigami principles have also driven material science advancements in parachute design. Researchers are exploring:\n- Lightweight composite materials\n- Shape-memory alloys\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "- Smart fabrics that can change properties\n- Integrated sensors for real-time performance monitoring\n\nThese material innovations complement the cutting and folding techniques,",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                " creating parachutes that are not just structural devices but intelligent systems.\n\nPractical Applications\n\nThe potential applications for kirigami-inspired parachutes are vast and diverse:\n\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "Space Exploration\nNASA and other space agencies are particularly interested in these designs for planetary landing missions. The ability to create compact,",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                " adaptable parachutes could revolutionize how scientific payloads are delivered to extraterrestrial surfaces.\n\nDisaster Response\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "Emergency response teams could benefit from parachutes that can be easily transported and quickly deployed in challenging terrain or during complex rescue operations.\n\nMilitary and Defense\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "Precision payload delivery, reconnaissance equipment deployment, and emergency escape systems could be significantly improved using kirigami-inspired techniques.\n\nScientific Research\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "Scientific Research (Continued)\n\nAtmospheric and environmental researchers can utilize these advanced parachutes for:\n- Collecting atmospheric samples\n- Monitoring climate conditions\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "- Deploying scientific instruments in remote or challenging environments\n- Conducting long-duration aerial observations\n\nCommercial and Consumer Applications\n- Drone delivery systems\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "- Recreational sports equipment\n- Compact emergency safety devices\n- Precision agriculture monitoring tools\n\nChallenges and Ongoing Research\n\nDespite the promising potential,",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                " kirigami-inspired parachute designs face several challenges:\n\n1. Computational Modeling\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "Accurately predicting the behavior of complex kirigami-based structures requires advanced computational techniques. Researchers are developing sophisticated simulation models to understand:\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "- Stress distribution\n- Fluid dynamics\n- Material deformation\n- Performance under various environmental conditions\n\n2. Manufacturing Complexity\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "Creating precisely cut and folded parachute structures demands:\n- Advanced manufacturing techniques\n- High-precision cutting technologies\n- Specialized materials\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "- Sophisticated quality control processes\n\n3. Material Limitations\nFinding materials that can:\n- Withstand repeated folding\n- Maintain structural integrity\n- Remain lightweight\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "- Resist environmental degradation\n\nCutting-Edge Research Institutions\n\nSeveral research institutions are at the forefront of kirigami-inspired parachute development:\n\n1.",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                " MIT's Morphing Mechanisms Laboratory\nPioneering research in adaptive structural design, exploring how kirigami principles can be applied to aerospace engineering.\n\n2. NASA Jet Propulsion Laboratory\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "Developing kirigami-inspired deployment mechanisms for planetary exploration missions.\n\n3. Stanford University's Mechanical Engineering Department\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "Investigating advanced material technologies and computational modeling for kirigami structures.\n\nFuture Outlook\n\nThe future of kirigami-inspired parachutes looks incredibly promising.",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                " Anticipated developments include:\n\nShort-Term Predictions (5-10 years)\n- Refined computational modeling\n- Improved manufacturing techniques\n- Initial practical applications in specialized fields\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "- Increased material science innovations\n\nLong-Term Potential (10-20 years)\n- Fully adaptive, intelligent parachute systems\n- Seamless integration with autonomous technologies\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "- Revolutionary approaches to aerial payload delivery\n- Potential breakthroughs in multiple scientific and commercial domains\n\nInterdisciplinary Collaboration\n\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "The development of kirigami-inspired parachutes represents a perfect example of interdisciplinary innovation, requiring collaboration between:\n- Aerospace engineers\n- Materials scientists\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "- Computer scientists\n- Mathematical modelers\n- Design engineers\n- Computational fluid dynamicists\n\nConclusion\n\n",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                "Kirigami-inspired parachutes represent more than just a technological improvement—they signify a fundamental reimagining of how we approach aerial deployment systems.",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                " By drawing inspiration from an ancient Japanese art form and combining it with cutting-edge engineering principles, researchers are creating solutions that are lighter, more adaptable,",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                " and more intelligent than ever before.\n\nThe journey from paper-cutting art to advanced aerospace technology demonstrates the incredible potential of cross-disciplinary thinking.",
            ),
        ),
    ),
    Ok(
        ChatChunk(
            Content(
                " As research continues, we can expect kirigami-inspired designs to play an increasingly significant role in how we interact with and explore our world from above.\n\n",
            ),
        ),
    ),
    Ok(
        EndOfStream(
            Completed,
        ),
    ),
]
